doc/man/fedfs-delete-junction.8
.\"@(#)fedfs-delete-junction.8"
.\"
.\" @file doc/man/fedfs-delete-junction.8
.\" @brief man page for fedfs-delete-junction client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-DELETE-JUNCTION 8 "@publication-date@"
.SH NAME
fedfs-delete-junction \- send a FEDFS_DELETE_JUNCTION ADMIN protocol request
.SH SYNOPSIS
.B fedfs-delete-junction
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.I path
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-delete-junction (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_DELETE_JUNCTION request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_DELETE_JUNCTION request deletes a FedFS junction
in a local file system on a remote file server.
The
.BR fedfs-delete-junction (8)
command takes a single positional parameter which is the
pathname on the remote server of the junction to be deleted.
This pathname is relative to the root
of the local file system on the remote server.
.P
The FEDFS_DELETE_JUNCTION request does not remove the FSN record
associated with the deleted junction.
Other junctions may continue to refer to the FSN in the deleted junction.
.P
When no junction refers to this FSN, use the
.BR nsdb-delete-fsn (8)
command to delete the FSN and children FSL records.
Resolving a junction that contains an FSN UUID
without a matching FSN record on the NSDB
results in an error on the file server.
.P
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-delete-junction (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The pathname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADNAME
A component of the pathname consisted of valid UTF-8 characters
supported by the remote server,
but the name is not supported by the remote server
as a valid name for the current operation.
.TP
.B FEDFS_ERR_NAMETOOLONG
The length of the pathname exceeds the remote server’s implementation limit.
.TP
.B FEDFS_ERR_LOOP
Too many symbolic links were encountered in resolving the pathname.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_NOSPC
The requested operation would have caused the remote
server’s filesystem to exceed some limit (for example, if there is
a fixed number of junctions per fileset or per server).
.TP
.B FEDFS_ERR_NOTJUNCT
The pathname does not end in a junction.
.TP
.B FEDFS_ERR_NOTLOCAL
The pathname contains a junction in any position other than the last component.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_ROFS
A modifying operation was attempted on a read-only filesystem.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_PATH_TYPE_UNSUPP
The remote server does not support the specified FedFsPathType value.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain.
To delete the existing FedFS junction
.I /export/junction1
on the file server
.IR fs.example.net ,
use:
.RS
.sp
$ fedfs-delete-junction -h fs.example.net /export/junction1
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/fedfsc/fedfs-delete-junction.c
/**
 * @file src/fedfsc/fedfs-delete-junction.c
 * @brief Send a FEDFS_DELETE_JUNCTION RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <rpc/clnt.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Default RPC request timeout
 */
static struct timeval fedfs_delete_junction_timeout = { 25, 0 };

/**
 * Short form command line options
 */
static const char fedfs_delete_junction_opts[] = "?dh:n:";

/**
 * Long form command line options
 */
static const struct option fedfs_delete_junction_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nettype", 1, NULL, 'n', },
	{ NULL, 0, NULL, 0, },
};

static void
fedfs_delete_junction_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"path\n\n", progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

static FedFsStatus
fedfs_delete_junction_call(const char *hostname, const char *nettype,
		const char *path)
{
	enum clnt_stat status;
	FedFsStatus result;
	char **path_array;
	CLIENT *client;
	FedFsPath arg;

	memset(&arg, 0, sizeof(arg));

	result = nsdb_posix_to_path_array(path, &path_array);
	if (result != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result));
		return result;
	}
	result = nsdb_path_array_to_fedfspathname(path_array,
					&arg.FedFsPath_u.adminPath);
	if (result != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result));
		nsdb_free_string_array(path_array);
		return result;
	}

	client = clnt_create(hostname, FEDFS_PROG, FEDFS_V1, nettype);
	if (client == NULL) {
		clnt_pcreateerror("Failed to create FEDFS client");
		result = FEDFS_ERR_SVRFAULT;
		goto out;
	}

	memset((char *)&result, 0, sizeof(result));
	status = clnt_call(client, FEDFS_DELETE_JUNCTION,
				(xdrproc_t)xdr_FedFsPath, (caddr_t)&arg,
				(xdrproc_t)xdr_FedFsStatus, (caddr_t)&result,
				fedfs_delete_junction_timeout);
	if (status != RPC_SUCCESS) {
		clnt_perror(client, "FEDFS_DELETE_JUNCTION call failed");
		result = FEDFS_ERR_SVRFAULT;
	} else
		nsdb_print_fedfsstatus(result);
	(void)clnt_destroy(client);

out:
	nsdb_free_fedfspathname(&arg.FedFsPath_u.adminPath);
	nsdb_free_string_array(path_array);
	return result;
}

int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype, *path;
	unsigned int seconds;
	FedFsStatus status;
	int arg;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	hostname = "localhost";
	nettype = "netpath";
	while ((arg = getopt_long(argc, argv, fedfs_delete_junction_opts, fedfs_delete_junction_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		default:
			fprintf(stderr, "Invalid command line argument: %c\n", (char)arg);
		case '?':
			fedfs_delete_junction_usage(progname);
		}
	}
	if (argc == optind + 1)
		path = argv[optind];
	else if (argc > optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		fedfs_delete_junction_usage(progname);
	} else {
		fprintf(stderr, "No junction pathname was specified\n");
		fedfs_delete_junction_usage(progname);
	}

	for (seconds = FEDFS_DELAY_MIN_SECS;; seconds = fedfs_delay(seconds)) {
		status = fedfs_delete_junction_call(hostname, nettype, path);
		if (status != FEDFS_ERR_DELAY)
			break;

		xlog(D_GENERAL, "Delaying %u seconds...", seconds);
		if (sleep(seconds) != 0)
			break;
	}
	return (int)status;
}
